### PR TITLE
Use RMM adaptor constructors instead of factories.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR uses RMM memory resource adaptor constructors instead of factory functions. With CTAD, we do not need the factory and can use the constructor directly. The factory will be deprecated in https://github.com/rapidsai/rmm/pull/1626.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: java/src/main/native/src/RmmJni.cpp [hunk 1]
```diff
@@ -154,13 +154,6 @@ class tracking_resource_adaptor final : public base_tracking_resource_adaptor {
   }
 };
 
-template <typename Upstream>
-tracking_resource_adaptor<Upstream>* make_tracking_adaptor(Upstream* upstream,
```
Comment (bdice) [inline java/src/main/native/src/RmmJni.cpp:158]: I saw no uses of this function in the `rapidsai` or `NVIDIA` orgs (including `spark-rapids-jni`), so it should be safe to remove.

File: cpp/include/cudf_test/stream_checking_resource_adaptor.hpp [hunk 2]
```diff
@@ -156,16 +156,4 @@ class stream_checking_resource_adaptor final : public rmm::mr::device_memory_res
                                // cudf::test::get_default_stream() is observed.
 };
 
-/**
- * @brief Convenience factory to return a `stream_checking_resource_adaptor` around the
- * upstream resource `upstream`.
- *
- * @param upstream Reference to the upstream resource
- */
-inline stream_checking_resource_adaptor make_stream_checking_resource_adaptor(
```
Comment (bdice) [inline cpp/include/cudf_test/stream_checking_resource_adaptor.hpp:165]: I fixed the one usage of this function to use a constructor, so it should be safe to delete.

**Review by ttnghia — APPROVED** [review]

**Review by JayjeetAtGithub — APPROVED** [review]
Looks good to me

File: cpp/benchmarks/fixture/benchmark_fixture.hpp [hunk 3]
```diff
@@ -107,7 +107,7 @@ class memory_stats_logger {
  public:
   memory_stats_logger()
     : existing_mr(rmm::mr::get_current_device_resource()),
-      statistics_mr(rmm::mr::make_statistics_adaptor(existing_mr))
+      statistics_mr(rmm::mr::statistics_adaptor(existing_mr))
```
Comment (bdice) [inline cpp/benchmarks/fixture/benchmark_fixture.hpp:110]:
```suggestion
      statistics_mr(rmm::mr::statistics_resource_adaptor(existing_mr))
```

**Pull request merged**